readthedocs/search/api.py

```diff
@@ -118,19 +157,26 @@ def get_serializer_context(self):
 
     def get_all_projects(self):
         """
-        Return a list containing the project itself and all its subprojects.
-
-        The project slug is retrieved from ``project`` query param.
+        Return a list containing the project itself and all its subprojects the user has permissions over.
 
         :rtype: list
-
-        :raises: Http404 if project is not found
         """
-        project_slug = self.request.query_params.get('project')
-        version_slug = self.request.query_params.get('version')
-        all_projects = utils.get_project_list_or_404(
-            project_slug=project_slug, user=self.request.user, version_slug=version_slug,
+        main_version = self._get_version()
+        main_project = self._get_project()
+
+        subprojects = Project.objects.filter(
+            superprojects__parent_id=main_project.id,
         )
+        all_projects = []
+        for project in list(subprojects) + [main_project]:
+            version = (
+                Version.internal
+                .public(user=self.request.user, project=project)
+                .filter(slug=main_version.slug)
+                .first()
+            )
+            if version:
+                all_projects.append(version.project)
         return all_projects
```

Comment:
We should probably rename this `all_versions`, since it's actually a list of versions, not projects :)

Reply:
It's a list of projects, version is used to validate the user has access to the version of that subproject before searching.